src/operations/lists.rs

```diff
@@ -31,65 +31,288 @@
 //! If an index is out of bounds, a parameter error will be returned. If a range is partially out of
 //! bounds, the valid part of the range will be returned.
 
-use crate::operations::cdt::{CdtArgument, CdtOpType, CdtOperation};
+use crate::msgpack::encoder::pack_cdt_op;
+use crate::operations::cdt::{CdtArgument, CdtOperation};
+use crate::operations::cdt_context::{CdtContext, DEFAULT_CTX};
 use crate::operations::{Operation, OperationBin, OperationData, OperationType};
 use crate::Value;
 
+#[derive(Debug, Clone, Copy)]
+#[doc(hidden)]
+pub enum CdtListOpType {
+    SetType = 0,
+    Append = 1,
+    AppendItems = 2,
+    Insert = 3,
+    InsertItems = 4,
+    Pop = 5,
+    PopRange = 6,
+    Remove = 7,
+    RemoveRange = 8,
+    Set = 9,
+    Trim = 10,
+    Clear = 11,
+    Increment = 12,
+    Sort = 13,
+    Size = 16,
+    Get = 17,
+    GetRange = 18,
+    GetByIndex = 19,
+    GetByRank = 21,
+    GetByValue = 22,
+    GetByValueList = 23,
+    GetByIndexRange = 24,
+    GetByValueInterval = 25,
+    GetByRankRange = 26,
+    GetByValueRelRankRange = 27,
+    RemoveByIndex = 32,
+    RemoveByRank = 34,
+    RemoveByValue = 35,
+    RemoveByValueList = 36,
+    RemoveByIndexRange = 37,
+    RemoveByValueInterval = 38,
+    RemoveByRankRange = 39,
+    RemoveByValueRelRankRange = 40,
+}
+
+/// List storage order.
+#[derive(Debug, Clone, Copy)]
+pub enum ListOrderType {
+    /// List is not ordered. This is the default.
+    Unordered = 0,
+    /// List is ordered.
+    Ordered = 1,
+}
+
+/// `CdtListReturnType` determines the returned values in CDT List operations.
+#[derive(Debug, Clone, Copy)]
+pub enum ListReturnType {
+    /// Do not return a result.
+    None = 0,
+    /// Return index offset order.
+    /// 0 = first key
+    /// N = Nth key
+    /// -1 = last key
+    Index = 1,
+    /// Return reverse index offset order.
+    /// 0 = last key
+    /// -1 = first key
+    ReverseIndex = 2,
+    /// Return value order.
+    /// 0 = smallest value
+    /// N = Nth smallest value
+    /// -1 = largest value
+    Rank = 3,
+    /// Return reserve value order.
+    /// 0 = largest value
+    /// N = Nth largest value
+    /// -1 = smallest value
+    ReverseRank = 4,
+    /// Return count of items selected.
+    Count = 5,
+    /// Return value for single key read and value list for range read.
+    Values = 7,
+    /// Invert meaning of list command and return values.
+    /// With the INVERTED flag enabled, the items outside of the specified index range will be returned.
+    /// The meaning of the list command can also be inverted.
+    /// With the INVERTED flag enabled, the items outside of the specified index range will be removed and returned.
+    Inverted = 0x10000,
```

Comment:
The `Inverted` "return type" is another good example, how the format of the wire format bleeds through into the API design, leading to a client API that's much more difficult for developers to understand/use.

E.g. if I have a bin with a list of tags (e.g. "red", "green", "blue") and I want to remove all tags _but_ the "green" tag, I can use the `remove_by_value` operation. But I have to specify a `ListReturnType.None` with the `ListReturnType.Inverted` flag set to do so, which doesn't really make much sense. After all, I don't really care about the return values at all, but the `Inverted` flag also affects the `remove` operation itself!

It would be much nicer, if the list operation itself was its own type, that had an `invert_selection` method. When serializing the operation to the wire, the client would still combine the return type and the `inverted` flag into the same byte value, of course.

Reply:
Example again for how this looks like in the Node.js client:

```
  await client.put(key, { tags: ['red', 'green', 'orange', 'blue'] })
  const result = await client.operate(key, [
    lists.removeByValue('tags', 'green')
      .invertSelection()
      .andReturn(returnType.VALUE)
  ])
  console.log(result.bins) // => { tags: [ 'red', 'orange', 'blue' ] }
  const record = await client.get(key)
  console.log(record.bins) // => { tags: [ 'green' ] }
```

The downside, of course, is that there is additional over-head for the extra function calls to construct the operation, as well as potentially the additional memory required to keep the return type and the inverted flag in two separate fields of the list operation struct, instead of combining them.

I think it's a worthwhile trade-off for a "nicer" client API. But would be best to run some micro-benchmarks to get a better sense of the actual performance overhead.

Reply:
Ah ok, i get it. I agree it would make more sense to split that up. Now that its not the first time, the Operation <-> CdtOperation problem appeared. Maybe we should target a change of that before a 1.0 release. I would not like to do it in this PR, because this will probably become a similar large change and would be even harder to review.

Reply:
Agree. We should proceed with this PR as is, and do any refactoring (if we decided to do so) in a separate PR.